Make masked_fill reference traceable

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92380
* #90967
* #90887
* __->__ #90972

As the comment states, `item()` cannot be used since you can't trace through a
scalar.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

peterbell10 committed:
Make masked_fill reference traceable

As the comment states, `item()` cannot be used since you can't trace through a
scalar.

[ghstack-poisoned]

peterbell10 committed:
Update on "Make masked_fill reference traceable"


As the comment states, `item()` cannot be used since you can't trace through a
scalar.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "Make masked_fill reference traceable"


As the comment states, `item()` cannot be used since you can't trace through a
scalar.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "Make masked_fill reference traceable"


As the comment states, `item()` cannot be used since you can't trace through a
scalar.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -4728,21 +4728,23 @@ def masked_fill(a: TensorLikeType, mask: TensorLikeType, value: TensorOrNumberLi
+        if a.device.type == "cuda" and value.device.type == "cpu":
+            # NOTE: Could not use value = item(value) as it resulted in
+            # RuntimeError: Cannot cast FakeTensor(cpu) to number
+            value = prims.device_put(value, a.device)
```
I think this is no longer needed, as `where` accepts 0d cpu tensors

peterbell10 replied:
I tried this and got 
> FAILED test/inductor/test_torchinductor.py::CudaTests::test_masked_fill_promotion_cuda - RuntimeError: CUDA error: an illegal memory access was encountered

ngimel replied:
Ok looks like `where` inductor lowering lacks those device_put calls, we should fix that, but `masked_fill` can stay like this for now.

ngimel replied:
Turns out `where` lowering in inductor is fine, but inductor doesn't support `prims.to_dtype` on line 4765. Replacing it with `prims.convert_element_type` fixes test error even without `device_put` and makes generated code much nicer (with this decomp it falls back to prims.to_dtype which we definitely want to avoid, and generates a cpu kernel for a 0d cpu tensor, which is also very wasteful). 
@mruberry why do we have both `convert_element_type` and `to_dtype` and what's the difference between them? Should we remove `to_dtype`, its meta even explicitly says it's incorrect.

mruberry replied:
Yes, removing `to_dtype` makes sense. It was a temporary addition to explore stride propagation

ngimel replied:
Right, `_maybe_convert_to_dtype` is more appropriate, as conversion is not always needed.

peterbell10 committed:
Update on "Make masked_fill reference traceable"


As the comment states, `item()` cannot be used since you can't trace through a
scalar.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "Make masked_fill reference traceable"


As the comment states, `item()` cannot be used since you can't trace through a
scalar.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "Make masked_fill reference traceable"


As the comment states, `item()` cannot be used since you can't trace through a
scalar.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "Make masked_fill reference traceable"


As the comment states, `item()` cannot be used since you can't trace through a
scalar.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "Make masked_fill reference traceable"


As the comment states, `item()` cannot be used since you can't trace through a
scalar.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "Make masked_fill reference traceable"


As the comment states, `item()` cannot be used since you can't trace through a
scalar.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "Make masked_fill reference traceable"


As the comment states, `item()` cannot be used since you can't trace through a
scalar.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "Make masked_fill reference traceable"


As the comment states, `item()` cannot be used since you can't trace through a
scalar.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge -f "dynamo failure is pre-existing"
